File: python/523.py
"""Project Euler Problem 523: First Sort I.

Consider a sorting algorithm where we repeatedly take the first pair of
adjacent elements not in order, and move the smaller element to the
beginning of the list. Find the expected number of moves for this
algorithm over all permutations of N elements.
"""

from __future__ import annotations


def solve() -> str:
    """Solve Problem 523."""
    N = 30
    ans = 0.0

    for n in range(2, N + 1):
        for i in range(n - 1):
            ans += (2**i) / n

    return f"{ans:.2f}"


def main() -> str:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/523.c
/*
 * Project Euler 523 - First Sort I
 *
 * Expected number of moves for the sorting algorithm over all
 * permutations of N elements. Formula: sum over n=2..N of
 * sum over i=0..n-2 of 2^i / n.
 */
#include <stdio.h>
#include <math.h>

int main(void) {
    int N = 30;
    double ans = 0.0;

    for (int n = 2; n <= N; n++) {
        double pow2 = 1.0;
        for (int i = 0; i < n - 1; i++) {
            ans += pow2 / n;
            pow2 *= 2.0;
        }
    }

    printf("%.2f\n", ans);
    return 0;
}
